extensions/chatgpt/src/views/vision.tsx

```diff
@@ -0,0 +1,182 @@
+      const imageHeight = data.type.height;
+      setImageMeta({ height: imageHeight, width: imageWidth, size: data.data.length });
+      setImagePromptTokenCount(countImageTokens(imageWidth, imageHeight));
+      setPromptTokenCount(countToken(VISION_MODEL.prompt + prompt));
+      return streamOrCompletion;
+    } catch (error) {
+      await showToast({ style: Toast.Style.Failure, title: "Error" });
+      setLoading(false);
+      setResponse(
+        "## ⚠️ Issue when accessing the API. \n\n" + `Error Message: \n\n \`\`\`${(error as Error).message}\`\`\``
+      );
+      return;
+    }
+  }
+
+  async function getResult() {
+    const now = new Date();
+    let duration = 0;
+    const toast = await showToast(Toast.Style.Animated, toast_title);
+
+    const resp = await getChatResponse(user_prompt);
+    if (!resp) return;
+
+    let response_ = "";
+    function appendResponse(part: string) {
+      response_ += part;
+      setResponse(response_);
+      setResponseTokenCount(countToken(response_));
+    }
+
+    if (resp instanceof Stream) {
+      for await (const part of resp) {
+        appendResponse(part.choices[0]?.delta?.content ?? "");
+      }
+    } else {
+      appendResponse(resp.choices[0]?.message?.content ?? "");
+    }
+
+    setLoading(false);
+    const done = new Date();
+    duration = (done.getTime() - now.getTime()) / 1000;
+    toast.style = Toast.Style.Success;
+    toast.title = `Finished in ${duration} seconds`;
+  }
+
+  useEffect(() => {
+    getResult();
+  }, []);
+
+  useEffect(() => {
+    if (loading == false) {
+      setCumulativeTokens(cumulative_tokens + prompt_token_count + response_token_count + image_prompt_token_count);
+      setCumulativeCost(
+        cumulative_cost +
+          estimatePrice(prompt_token_count, response_token_count, VISION_MODEL.option) +
+          estimateImagePrice(image_prompt_token_count)
+      );
+    }
+  }, [loading]);
+
+  return (
+    <Detail
+      markdown={response}
+      isLoading={loading}
+      actions={
+        !loading && (
+          <ActionPanel title="Actions">
+            <Action.CopyToClipboard title="Copy Results" content={response} />
+            <Action.Paste title="Paste Results" content={response} />
+          </ActionPanel>
+        )
+      }
+      metadata={
+        <Detail.Metadata>
+          <Detail.Metadata.Separator />
+          {imageMeta.size && <Detail.Metadata.Label title="Size" text={toUnit(imageMeta.size)} />}
+          {imageMeta.width && <Detail.Metadata.Label title="Width" text={String(imageMeta.width)} />}
+          {imageMeta.height && <Detail.Metadata.Label title="Height" text={String(imageMeta.height)} />}
+
+          <Detail.Metadata.Separator />
+          <Detail.Metadata.Label title="Culmulative Tokens" text={cumulative_tokens.toString()} />
+          <Detail.Metadata.Label title="Culmulative Cost" text={cumulative_cost.toFixed(4) + " ¢"} />
+        </Detail.Metadata>
```

Comment:
Can we do something here so it doesn't look weird when there is an error?

![chatgpt 2024-06-20 at 11 45 36](https://github.com/raycast/extensions/assets/38033107/f3636515-a7e7-40b2-a324-71785b8f6a47)

Could we make `Culmulative token/cost` a preference? Would that make sense at all? I can't see that information anywhere else in this extension 🙂

Reply:
push -> hid metadata when error. is that ok?
I used only gpt-4o for vision as other models like gpt-3 (chatgpt default model) doesn't support.
cumulative token are only calculated on gpt-4o price.
`  switch (model) {
    case "gpt-4o":
      // The input costs are $5, while the output is $15.
      price = ((prompt_token * 5) / 1_000_000 + (output_token * 15) / 1_000_000) * 100;
      break;
    default:
      break;
  }`
 because users can add any models to it it can't estimate other models.

alternatively I can just remove the whole metadata thing